### PR TITLE
Allow websocket connection trough "Basic Authorization" headers

### PR DESCRIPTION
**Pull request by TalissonJunior**
I´m currently using angular as the client app with an implemented HTTP Interceptor to put the "Authorization" headers on each request, 
it would be great if we could connect to the WebSocket using the "Authorization" headers, 

besides.... my session is always getting null value for the "userId", which throws an parse error  on line
```
long userId = (Long) ((HttpSession) req.getSession()).getAttribute(SessionResource.USER_ID_KEY);
```
Allowing the connection trough the header would solve the problem.

### CODE REVIEW

**Comment by TalissonJunior** [comment]
I tried to make the websocket call ```SecurityRequestFilter.filter```but it doesn´t seems to work, and this function requires a parameter of ```ContainerRequestContext``` which is different from the websocket ```JettyServerUpgradeRequest```, even a cast would be a problem...

But an alternative would be moving the repeat code into a static function ```GetUserIdFromHeaders``,
so the full code would be like: 

SecurityRequestFilter
```
 @Override
    public void filter(ContainerRequestContext requestContext) {

        if (requestContext.getMethod().equals("OPTIONS")) {
            return;
        }

        SecurityContext securityContext = null;

        try {

            String authHeader = requestContext.getHeaderString(AUTHORIZATION_HEADER);

            if (authHeader != null) {
                try {
                    Long userId = GetUserIdFromHeaders(authHeader);

                    if(userId != null) {
                        Main.getInjector().getInstance(StatisticsManager.class).registerRequest(userId);
                        securityContext = new UserSecurityContext(new UserPrincipal(userId));
                    }
                } catch (StorageException e) {
                    throw new WebApplicationException(e);
                }
            } else if (request.getSession() != null) {

                Long userId = (Long) request.getSession().getAttribute(SessionResource.USER_ID_KEY);
                if (userId != null) {
                    Context.getPermissionsManager().checkUserEnabled(userId);
                    Main.getInjector().getInstance(StatisticsManager.class).registerRequest(userId);
                    securityContext = new UserSecurityContext(new UserPrincipal(userId));
                }

            }

        } catch (SecurityException e) {
            LOGGER.warn("Authentication error", e);
        }

        if (securityContext != null) {
            requestContext.setSecurityContext(securityContext);
        } else {
            Method method = resourceInfo.getResourceMethod();
            if (!method.isAnnotationPresent(PermitAll.class)) {
                Response.ResponseBuilder responseBuilder = Response.status(Response.Status.UNAUTHORIZED);
                if (!XML_HTTP_REQUEST.equals(request.getHeader(X_REQUESTED_WITH))) {
                    responseBuilder.header(WWW_AUTHENTICATE, BASIC_REALM);
                }
                throw new WebApplicationException(responseBuilder.build());
            }
        }

    }

    public static Long GetUserIdFromHeaders(String authHeaders) throws StorageException {
        if(authHeaders == null) {
            return null;
        }

        String[] auth = decodeBasicAuth(authHeaders);
        User user = Context.getPermissionsManager().login(auth[0], auth[1]);

        if (user != null) {
            return user.getId();
        }

        return null;
    }
```

AsyncSocketServlet
```
@Override
    public void configure(JettyWebSocketServletFactory factory) {
        factory.setIdleTimeout(Duration.ofMillis(Context.getConfig().getLong(Keys.WEB_TIMEOUT)));

        factory.setCreator((req, resp) -> {
            String authHeader = req.getHeader(AUTHORIZATION_HEADER);

            if (authHeader != null) {
                try {
                    Long userId = SecurityRequestFilter.GetUserIdFromHeaders(authHeader);

                    if(userId == null) {
                        return null;
                    }
                    return new AsyncSocket(userId);
                } catch (StorageException e) {
                    throw new WebApplicationException(e);
                }
            } else {
                if (req.getSession() != null) {
                    long userId = (Long) ((HttpSession) req.getSession()).getAttribute(SessionResource.USER_ID_KEY);
                    return new AsyncSocket(userId);
                } else {
                    return null;
                }
            }
        });
    }
```

I have tested, and it works the same way as the before, i can make a pull request for a better preview...

**Comment by TalissonJunior** [comment]
i see, well we could use a filter with a little different approach, since i coudn´t reach the "Authorization Headers" trough it

so basically on the client app we would connect to the socket using a query parameter 
ex: url "../api/socket?Authorization=MyBasicAuth",


Then the file ```FilterSocket```(New File) which contains 
```
 @Override
    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {

        HttpServletRequest httpRequest = (HttpServletRequest) request;
        String authHeaders = httpRequest.getQueryString();
        String queryAuthName = AUTHORIZATION_HEADER + "=";

        if (authHeaders != null && authHeaders.contains(queryAuthName)) {
            String[] auth = decodeBasicAuth(authHeaders.replace(queryAuthName, ""));
            if(auth != null && auth.length == 2) {

                try {
                    User user = Context.getPermissionsManager().login(auth[0], auth[1]);

                    if (user != null) {
                        httpRequest.getSession().setAttribute(SessionResource.USER_ID_KEY, user.getId());
                    }
                }
                catch (StorageException e) {}
            }
        }

        chain.doFilter(httpRequest, response);
    }
```
would get the id of the user based on the authorization and set it to the session, before calling the websocket  ```AsyncSocketServlet``` which would be the same as it was before

AsyncSocketServlet
```
 @Override
    public void configure(JettyWebSocketServletFactory factory) {
        factory.setIdleTimeout(Duration.ofMillis(Context.getConfig().getLong(Keys.WEB_TIMEOUT)));
        factory.setCreator((req, resp) -> {
            if (req.getSession() != null) {
                long userId = (Long) ((HttpSession) req.getSession()).getAttribute(SessionResource.USER_ID_KEY);
                return new AsyncSocket(userId);
            } else {
                return null;
            }
        });
    }
```

What do you think?
note: the query parameter is not required

**Comment by tananaev** [comment]
> since i coudn´t reach the "Authorization Headers" trough it

So, do you get a filter call for the websocket connection?

**Comment by tananaev** [comment]
I think you misunderstood me. I want to reuse existing filter.

**Comment by TalissonJunior** [comment]
i see, when with existing filter you mean "SecurityRequestFilter" right ?

**Comment by tananaev** [comment]
Correct.

**Comment by TalissonJunior** [comment]
ok i got it, then yeah, i removed the file that i created and implement the filter on the ```SecurityRequestFilter``` as the folowing code shows: 

```
 @Override
    public void filter(ContainerRequestContext requestContext) {

        if (requestContext.getMethod().equals("OPTIONS")) {
            return;
        }

        SecurityContext securityContext = null;

        try {

            String authHeader = requestContext.getHeaderString(AUTHORIZATION_HEADER);

            if (authHeader != null) {
                try {
                    Long userId = GetUserIdFromBasicAuth(authHeader);

                    if(userId != null) {
                        Main.getInjector().getInstance(StatisticsManager.class).registerRequest(userId);
                        securityContext = new UserSecurityContext(new UserPrincipal(userId));
                    }
                } catch (StorageException e) {
                    throw new WebApplicationException(e);
                }
            } else if (request.getSession() != null) {

                Long userId = (Long) request.getSession().getAttribute(SessionResource.USER_ID_KEY);
                if (userId != null) {
                    Context.getPermissionsManager().checkUserEnabled(userId);
                    Main.getInjector().getInstance(StatisticsManager.class).registerRequest(userId);
                    securityContext = new UserSecurityContext(new UserPrincipal(userId));
                }

            }

        } catch (SecurityException e) {
            LOGGER.warn("Authentication error", e);
        }

        if (securityContext != null) {
            requestContext.setSecurityContext(securityContext);
        } else {
            Method method = resourceInfo.getResourceMethod();
            if (!method.isAnnotationPresent(PermitAll.class)) {
                Response.ResponseBuilder responseBuilder = Response.status(Response.Status.UNAUTHORIZED);
                if (!XML_HTTP_REQUEST.equals(request.getHeader(X_REQUESTED_WITH))) {
                    responseBuilder.header(WWW_AUTHENTICATE, BASIC_REALM);
                }
                throw new WebApplicationException(responseBuilder.build());
            }
        }

    }

    @Override
    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
        HttpServletRequest httpRequest = (HttpServletRequest) request;
        String authHeaders = httpRequest.getQueryString();
        String queryAuthName = AUTHORIZATION_HEADER + "=";

        if (authHeaders != null && authHeaders.contains(queryAuthName)) {
            try {
                Long userId = GetUserIdFromBasicAuth(authHeaders.replace(queryAuthName, ""));

                if (userId != null) {
                    httpRequest.getSession().setAttribute(SessionResource.USER_ID_KEY, userId);
                }
            }
            catch (StorageException e) {}
        }

        chain.doFilter(httpRequest, response);
    }

    private Long GetUserIdFromBasicAuth(String authHeaders) throws StorageException {
        if(authHeaders == null) {
            return null;
        }

        String[] auth = decodeBasicAuth(authHeaders);
        User user = Context.getPermissionsManager().login(auth[0], auth[1]);

        if (user != null) {
            return user.getId();
        }

        return null;
    }
```

**Comment by tananaev** [comment]
You need to send a PR because I don't know what the difference is.

File: src/main/java/org/traccar/api/security/SecurityRequestFilter.java [hunk 1]
```diff
@@ -116,4 +117,38 @@ public void filter(ContainerRequestContext requestContext) {
 
     }
 
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+        HttpServletRequest httpRequest = (HttpServletRequest) request;
+        String authHeaders = httpRequest.getQueryString();
```
Comment (tananaev) [inline src/main/java/org/traccar/api/security/SecurityRequestFilter.java:123]: Why can't we use headers?

Comment (TalissonJunior) [reply src/main/java/org/traccar/api/security/SecurityRequestFilter.java:123]: it seems that browsers don´t support sending headers with websocket, you can only send "protocols" or attach a cookie before calling the websocket
```
var ws = new WebSocket("ws://example.com/service", "soap");
```

**Comment by tananaev** [comment]
What about your original solution? It didn't work? I'm reading the title and description of the PR and it doesn't make sense to me if it's not possible in the first place.

**Comment by TalissonJunior** [comment]
i changed to get the authorization from headers as you suggest, and it wasn´t working , until i type the url : "api/devices" on the browser and enter my credentials, now i´m getting the authorization from headers...

**Comment by tananaev** [comment]
I'm confused. Are you saying your original PR didn't use headers?

**Comment by TalissonJunior** [comment]
I´m saying that my original PR did use headers, but it shouldn´t since browsers doesn´t allow it, so in my opinion we should close this PR, or allow Basic authorization trough queryParameter when working with socket.

I tried to use "the basic authorization" because the logic to get the userId from the current session was not working on my case, i was getting "WebSocket connection to 'ws://localhost:8082/api/socket' failed"  (UserId from session was always null)

Allowing basic authorization trought queryParameters would make this problem go away.

**Comment by tananaev** [comment]
So, your original code didn't work?

Let's close it for now.

**Pull request closed without merging**